slack_notification/slack_notification.php

```diff
@@ -1,183 +1,183 @@
-    'short' => 'true'
-  ),
-  array(
-    'title' => 'View Dashboard',
-    'value' => '<https://dashboard.pantheon.io/sites/'. PANTHEON_SITE .'#'. PANTHEON_ENVIRONMENT .'/deploys|View Dashboard>',
-    'short' => 'true'
-  ),
-);
-
-// Customize the message based on the workflow type.  Note that slack_notification.php
-// must appear in your pantheon.yml for each workflow type you wish to send notifications on.
-switch($_POST['wf_type']) {
-  case 'deploy':
-    // Find out what tag we are on and get the annotation.
-    $deploy_tag = `git describe --tags`;
-    $deploy_message = $_POST['deploy_message'];
-
-    // Prepare the slack payload as per:
-    // https://api.slack.com/incoming-webhooks
-    $text = 'Deploy to the '. $_ENV['PANTHEON_ENVIRONMENT'];
-    $text .= ' environment of '. $_ENV['PANTHEON_SITE_NAME'] .' by '. $_POST['user_email'] .' complete!';
-    $text .= ' <https://dashboard.pantheon.io/sites/'. PANTHEON_SITE .'#'. PANTHEON_ENVIRONMENT .'/deploys|View Dashboard>';
-    // Build an array of fields to be rendered with Slack Attachments as a table
-    // attachment-style formatting:
-    // https://api.slack.com/docs/attachments
-    $fields[] = array(
-      'title' => 'Details',
-      'value' => $text,
-      'short' => 'false'
-    );
-    $fields[] = array(
-      'title' => 'Deploy Note',
-      'value' => $deploy_message,
-      'short' => 'false'
-    );  
-    break;
-
-  case 'sync_code':
-    // Get the committer, hash, and message for the most recent commit.
-    $committer = `git log -1 --pretty=%cn`;
-    $email = `git log -1 --pretty=%ce`;
-    $message = `git log -1 --pretty=%B`;
-    $hash = `git log -1 --pretty=%h`;
-
-    // Prepare the slack payload as per:
-    // https://api.slack.com/incoming-webhooks
-    $text = 'Code sync to the ' . $_ENV['PANTHEON_ENVIRONMENT'] . ' environment of ' . $_ENV['PANTHEON_SITE_NAME'] . ' by ' . $_POST['user_email'] . "!\n";
-    $text .= 'Most recent commit: ' . rtrim($hash) . ' by ' . rtrim($committer) . ': ' . $message;
-    // Build an array of fields to be rendered with Slack Attachments as a table
-    // attachment-style formatting:
-    // https://api.slack.com/docs/attachments
-    $fields = array_merge($fields, array(
-      array(
-        'title' => 'Commit',
-        'value' => rtrim($hash),
-        'short' => 'true'
-      ),
-      array(
-        'title' => 'Commit Message',
-        'value' => rtrim($message),
-        'short' => 'false'
-      )
-    ));
-    break;
-
-  case 'clear_cache':
-    $fields[] = array(
-      'title' => 'Cleared caches',
-      'value' => 'Cleared caches on the ' . $_ENV['PANTHEON_ENVIRONMENT'] . ' environment of ' . $_ENV['PANTHEON_SITE_NAME'] . "!\n",
-      'short' => 'false'
-    );
-    break;
-
-  default:
-    $text = $_POST['qs_description'];
-    break;
+/**
+ * class Slack_Text - a basic { type, text } object to embed in a block
+ */
+class Slack_Text {
```

Comment:
This is kind of an awkward design. The lack of inheritance or methods suggests that classes are not the right abstraction. Appending classes via `new` in an array is an atypical API, and there are no seatbelts here to prevent the user from including the wrong type for a given object.

If you wanted to make a message factory, I'd recommend methods of a factory class over individual classes for each method type. However, note that even the [top result for php slack api](https://github.com/jolicode/slack-php-api/blob/main/docs/examples/posting-message.php) just uses json_encode for messages. I'd encourage you to keep the example simpler.

Reply:
This was a large part of the contributed PR (#176). My original design was more straightforward (but lacked extensibility), but perhaps these classes could just be functions. As a rule, I don't like multiple classes in a single file, but I was willing to overlook that in an effort to not complicate the script too much. Maybe I can find a happy medium.